vllm/v1/executor/multiproc_executor.py

```diff
@@ -0,0 +1,197 @@
+        if world_size > 1:
+            maybe_set_triton_cache_manager()
+
+        # Multiprocessing-based executor does not support multi-node setting.
+        # Since it only works for single node, we can use the loopback address
+        # 127.0.0.1 for communication.
+        distributed_init_method = get_distributed_init_method(
+            "127.0.0.1", get_open_port())
+
+        # Initialize worker and set up message queues for SchedulerOutputs
+        # and ModelRunnerOutputs
+        self.worker_request_mq = MessageQueue(world_size, world_size)
+        scheduler_output_handle = self.worker_request_mq.export_handle()
+
+        # Create workers
+        self.workers: List[WorkerProcHandle] = []
+        for rank in range(world_size):
+            worker = WorkerProc.make_worker_process(vllm_config, rank, rank,
+                                                    distributed_init_method,
+                                                    scheduler_output_handle)
+            self.workers.append(worker)
+
+        model_output_mq_handle = self.workers[0].model_output_mq_handle
+        self.model_output_mq = MessageQueue.create_from_handle(
+            model_output_mq_handle, 0)
+
+    def _run_on_workers(self, fn: str, *args) -> List:
+        with ThreadPoolExecutor() as executor:
+            futures = [
+                executor.submit(getattr(type(w), fn), w, *args)
+                for w in self.workers
+            ]
+            result = [f.result() for f in futures]  # Wait for all to complete
+        return result
+
+    def initialize(self, num_gpu_blocks: int) -> None:
+        """
+        Initialize the KV caches and begin the model execution loop of the
+        underlying workers.
+        """
+        success_vals = self._run_on_workers('initialize', num_gpu_blocks)
+        if not all(success_vals):
+            raise RuntimeError("Worker initialization failed.")
+
+        self.worker_request_mq.wait_until_ready()
+        self.model_output_mq.wait_until_ready()
+
+    def determine_num_available_blocks(self) -> Tuple[int, int]:
+        """
+        Determine the number of available KV blocks by invoking the
+        underlying worker.
+        """
+        # Get the maximum number of blocks that can be allocated on GPU and CPU.
+        num_blocks = self._run_on_workers('determine_num_available_blocks')
+
+        # Since we use a shared centralized controller, we take the minimum
+        # number of blocks across all workers to make sure all the memory
+        # operators can be applied to all workers.
+        num_gpu_blocks = min(b[0] for b in num_blocks)
+        num_cpu_blocks = min(b[1] for b in num_blocks)
+
+        return num_gpu_blocks, num_cpu_blocks
+
+    def execute_model(
+        self,
+        scheduler_output,
+    ) -> ModelRunnerOutput:
+        self.worker_request_mq.enqueue(scheduler_output)
+        model_output = self.model_output_mq.dequeue()
```

Comment:
This is a blocking call right?

Reply:
Yes and yes -- check out shm_broadcast.py. Had to sidestep an issue there to make sure `enqueue` and `dequeue` release the GIL

Reply:
okay I see now how the `acquire_read` and `acquire_write` are used in this pathway. Nice find on the time.sleep()!